### PR TITLE
[RDY] Fix the build for some 'clang' compilers [Vol 2]

### LINKED ISSUE #17087: Build error using clang/SDL on Mac OSX
Getting this when trying to build on a recent clone:

Macintosh-2:Cataclysm-DDA_tiles admin$ git pull && make NATIVE=osx OSX_MIN=10.7 RELEASE=1 LOCALIZE=0 CLANG=1 FRAMEWORK=1 TILES=1
remote: Counting objects: 33, done.
remote: Total 33 (delta 24), reused 24 (delta 24), pack-reused 9
Unpacking objects: 100% (33/33), done.
From git://github.com/CleverRaven/Cataclysm-DDA
   f9fa31f..dffc5a5  master     -> origin/master
Updating f9fa31f..dffc5a5
Fast-forward
 data/json/items/tools.json           |   7 ++--
 data/json/recipes/other/tool.json    |   4 +-
 data/json/recipes/recipe_ammo.json   |  10 ++---
 data/json/recipes/recipe_armor.json  |  55 ++++++++++++-------------
 data/json/recipes/recipe_others.json | 201 +++++++++++++++++++++++++++++++++----------------------------------------------------------
 data/json/recipes/recipe_weapon.json | 338 ++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------------------------------------------------------------
 data/json/tool_qualities.json        |   4 ++
 src/vehicle.cpp                      |   4 +-
 8 files changed, 232 insertions(+), 391 deletions(-)
/bin/sh: astyle: command not found
#define VERSION "0.C-15013-gdffc5a5"
Cannot run an astyle check, your system either does not have astyle, or it is too old.
tools/json_format_check.sh
clang++  -DRELEASE -DMACOSX -DGIT_VERSION -DTILES -ffast-math -Os  -Wall -Wextra   -Werror -stdlib=libc++ -std=c++11 -MMD -mmacosx-version-min=10.7 -F/Library/Frameworks -I/Library/Frameworks/SDL2.framework/Headers -I/Library/Frameworks/SDL2_image.framework/Headers -I/Library/Frameworks/SDL2_ttf.framework/Headers -c src/item.cpp -o obj/tiles/item.o
src/item.cpp:4284:30: error: chosen constructor is explicit in copy-initialization
    return { "", nullptr, 0, {} };
                             ^~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/set:428:14: note: constructor declared here
    explicit set(const value_compare& __comp = value_compare())

### PR DESCRIPTION
**Pull request by codemime**
Fixes #17087. To some extent this continues #16886 (the issue is quite similar).

### CODE REVIEW

**Comment by bpwatts** [comment]
Bless you a thousand times. Will test this when I get home.

File: src/item.h [hunk 1]
```diff
@@ -1229,6 +1229,15 @@ class item : public JsonSerializer, public JsonDeserializer, public visitable<it
             /** flags change behavior of gun mode and are **not** equivalent to item flags */
             std::set<std::string> flags;
 
+            gun_mode( const std::string &mode = "", item *target = nullptr, int qty = 0 ) :
```
Comment (mugling) [inline src/item.h]: These should be provided as default values for members themselves. Two constructors additional constructors to satisfy a compiler bug seems excessive

**Comment by mugling** [comment]
Which version of clang?

**Comment by codemime** [comment]
@bpwatts Does it work as expected?

**Comment by bpwatts** [comment]
I was literally about to check. Give me 5 minutes.

Macintosh-2:Cataclysm-DDA_tiles admin$ clang --version
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin13.4.0
Thread model: posix

File: src/item.h (hunk 1 above)
Comment (codemime) [inline src/item.h]: Good point, but at least one constructor has to be here to mend [this](https://github.com/CleverRaven/Cataclysm-DDA/pull/17098/files#diff-f7ca61cf749eba039da961e8a7be1630L4260) line (I expect it has the same problem).

**Comment by bpwatts** [comment]
No :(

src/item_factory.cpp:137:113: error: chosen constructor is explicit in copy-initialization
            obj.gun->modes.emplace( "DEFAULT", std::tuple<std::string, int, std::set<std::string>>( defmode, 1, {} ) );
                                                                                                                ^~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/set:428:14: note: constructor declared here
    explicit set(const value_compare& __comp = value_compare())
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/tuple:507:35: note: passing argument to parameter '__t' here
    explicit tuple(const _Tp& ... __t) _NOEXCEPT_((__all<is_nothrow_copy_constructible<_Tp>::value...>::value)) 
                                  ^
src/item_factory.cpp:141:126: error: chosen constructor is explicit in copy-initialization
                obj.gun->modes.emplace( "AUTO", std::tuple<std::string, int, std::set<std::string>>( "auto", obj.gun->burst, {} ) );
                                                                                                                             ^~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/set:428:14: note: constructor declared here
    explicit set(const value_compare& __comp = value_compare())
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/tuple:507:35: note: passing argument to parameter '__t' here
    explicit tuple(const _Tp& ... __t) _NOEXCEPT_((__all<is_nothrow_copy_constructible<_Tp>::value...>::value)) 
                                  ^
2 errors generated.
make: *** [obj/tiles/item_factory.o] Error 1

**Comment by codemime** [comment]
@bpwatts Thank you for the testing. At least the existing fixes do work, but there are more places that require mending.

**Comment by codemime** [comment]
@bpwatts I've fixed the reported errors. I'd appreciate if you could test this again.

**Comment by bpwatts** [comment]
I surely can! I think I have to download these individually, right?

**Comment by codemime** [comment]
You can fetch the whole `fix_clang_build_2` branch.

**Comment by bpwatts** [comment]
clang++  -DRELEASE -DMACOSX -DGIT_VERSION -DTILES -ffast-math -Os  -Wall -Wextra   -Werror -stdlib=libc++ -std=c++11 -MMD -mmacosx-version-min=10.7 -F/Library/Frameworks -I/Library/Frameworks/SDL2.framework/Headers -I/Library/Frameworks/SDL2_image.framework/Headers -I/Library/Frameworks/SDL2_ttf.framework/Headers -c src/item.cpp -o obj/tiles/item.o
src/item.cpp:4284:30: error: chosen constructor is explicit in copy-initialization
    return { "", nullptr, 0, {} };
                             ^~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/set:428:14: note: constructor declared here
    explicit set(const value_compare& __comp = value_compare())
             ^
src/item.h:1233:100: note: passing argument to parameter 'flags' here
            gun_mode( const std::string &mode, item *target, int qty, const std::set<std::string> &flags ) :
                                                                                                   ^
1 error generated.
make: *** [obj/tiles/item.o] Error 1

**Comment by codemime** [comment]
It appears that you're testing the `master` branch (prior to the changes).

**Comment by bpwatts** [comment]
Argh damnit I dont git gud. How would I fetch just that branch?

File: src/item.h [hunk 2]
```diff
@@ -1222,13 +1222,20 @@ class item : public JsonSerializer, public JsonDeserializer, public visitable<it
...
+            std::string mode;           /** name of this mode */
+            item *target = nullptr;     /** pointer to base gun or attached gunmod */
+            int qty = 0;                /** burst size or melee range */
 
             /** flags change behavior of gun mode and are **not** equivalent to item flags */
             std::set<std::string> flags;
 
+            gun_mode() = default;
+            gun_mode( const std::string &mode, item *target, int qty, const std::set<std::string> &flags ) :
+                mode( mode ),
```
Comment (mugling) [inline src/item.h]: better formatted on a single line

File: src/item_factory.cpp [hunk 3]
```diff
@@ -134,11 +134,13 @@ void Item_factory::finalize() {
             }
 
             // if the gun doesn't have a DEFAULT mode then add one now
-            obj.gun->modes.emplace( "DEFAULT", std::tuple<std::string, int, std::set<std::string>>( defmode, 1, {} ) );
+            obj.gun->modes.emplace( "DEFAULT", std::tuple<std::string, int, std::set<std::string>>( defmode, 1,
+                                    std::set<std::string>() ) );
```
Comment (mugling) [inline src/item_factory.cpp]: Ugly but I don't have a better proposal. What is clangs specific problem?

**Comment by bpwatts** [comment]
I'll tell you in a second, downloading mug's branch because I am illgitliterate

**Comment by codemime** [comment]
```
git fetch https://github.com/codemime/Cataclysm-DDA.git fix_clang_build_2
git checkout fix_clang_build_2
```

File: src/item.h (hunk 2 above)
Comment (codemime) [inline src/item.h]: Arguable. This formatting is both more readable and easier to modify.

File: src/item_factory.cpp (hunk 3 above)
Comment (codemime) [inline src/item_factory.cpp]: It's mentioned in the compiler's output: `chosen constructor is explicit in copy-initialization`. Old versions of `clang` seem to unable to deal with initializers.

**Comment by bpwatts** [comment]
it looks like your download worked, well past the compile point where it failed! Yup, it worked!

File: src/item.h (hunk 2 above)
Comment (mugling) [inline src/item.h]: The former is subjective. The latter how?

Comment (codemime) [inline src/item.h]: Initializer lists (obviously) initialize member variables, which are defined one per line. The list items and the variables should match (both order and usually number too). If you want to make sure (visually) they're in the right order, or reorder them, that's simply easier when they formatted the same way. That's enough for me to state the former and the latter.

File: src/item_factory.cpp (hunk 3 above)
Comment (codemime) [inline src/item_factory.cpp]: Possible solution for mitigating ugliness would be abandoning `tuple`s (and `pair`s). They're quite convenient for templates, but if templates aren't used here, why use `tuple`s? Explicit `struct` is much cleaner than a bunch of "anonymous" fields.

**Pull request merged**